pkg/jaegerquery/trace_query.go

```diff
@@ -0,0 +1,450 @@
+	   	ps_trace.jsonb(s.resource_tags) 	resource_tags,
+	   	ps_trace.jsonb(s.span_tags) 		span_tags,
+	   	array_agg(e.name) FILTER(WHERE e IS NOT NULL)			event_names,
+       		array_agg(e.time) FILTER(WHERE e IS NOT NULL)			event_times,
+	   	array_agg(e.dropped_tags_count) FILTER(WHERE e IS NOT NULL)	event_dropped_tags_count,
+	   	jsonb_agg(ps_trace.jsonb(e.tags)) FILTER(WHERE e IS NOT NULL)	event_tags,
+	   	inst_lib.name 				library_name,
+	   	inst_lib.version 			library_version,
+		inst_lib_url.url 			library_schema_url,
+		array_agg(lk.linked_trace_id) FILTER(WHERE lk IS NOT NULL)	links_linked_trace_ids,
+		array_agg(lk.linked_span_id)  FILTER(WHERE lk IS NOT NULL)	links_linked_span_ids,
+		array_agg(lk.trace_state)     FILTER(WHERE lk IS NOT NULL)	links_trace_states,
+		array_agg(lk.dropped_tags_count) FILTER(WHERE lk IS NOT NULL) 	links_dropped_tags_count,
+		jsonb_agg(lk.tags) FILTER(WHERE lk IS NOT NULL)			links_tags
+	FROM
+		_ps_trace.span s
+	INNER JOIN
+		_ps_trace.span_name sn ON s.name_id = sn.id
+	LEFT JOIN
+		_ps_trace.schema_url s_url ON s.resource_schema_url_id = s_url.id
+	LEFT JOIN
+		_ps_trace.event e ON e.span_id = s.span_id AND e.trace_id = s.trace_id
+	LEFT JOIN
+		_ps_trace.instrumentation_lib inst_lib ON s.instrumentation_lib_id = inst_lib.id
+	LEFT JOIN
+		_ps_trace.schema_url inst_lib_url ON inst_lib_url.id = inst_lib.schema_url_id
+	LEFT JOIN
+		_ps_trace.link lk ON lk.trace_id = s.trace_id AND lk.span_id = s.span_id
+	WHERE
+	  %s
+	GROUP BY
+	  s.trace_id,
+	  s.span_id,
+	  s.parent_span_id,
+	  s.start_time,
+	  s.end_time,
+	  s.resource_tags,
+	  s.span_tags,
+	  sn.name,
+	  s_url.url,
+	  inst_lib.name,
+	  inst_lib.version,
+	  inst_lib_url.url
+	%s`
+
+	subqueryFormat = `
+	SELECT
+		DISTINCT trace_id
+	FROM _ps_trace.span s
+	WHERE
+		%s
+	`
+)
+
+func buildTraceQuery(traceIDClause string, multipleTraces bool) string {
+	orderBy := ""
+	if multipleTraces {
+		orderBy = "ORDER BY s.trace_id"
+	}
+	return fmt.Sprintf(
+		fullTraceSQLFormat,
+		traceIDClause, orderBy)
+}
+
+func findTracesQuery(q *spanstore.TraceQueryParameters) (string, []interface{}) {
+	subquery, params := buildTraceIDSubquery(q)
+	traceIDClause := fmt.Sprintf("s.trace_id IN (%s)", subquery)
+	return buildTraceQuery(traceIDClause, true), params
+}
+
+func findTraceIDsQuery(q *spanstore.TraceQueryParameters) (string, []interface{}) {
+	subquery, params := buildTraceIDSubquery(q)
+	return subquery, params
+}
+
+func getTraceQuery(traceID model.TraceID) (string, []interface{}, error) {
+	var b [16]byte
+	n, err := traceID.MarshalTo(b[:])
+	if n != 16 || err != nil {
+		return "", nil, fmt.Errorf("marshaling TraceID: %w", err)
+	}
+
+	var uuid pgtype.UUID
+	uuid.Set(b)
+	params := []interface{}{uuid}
+
+	traceIDClause := "s.trace_id = $1"
+	return buildTraceQuery(traceIDClause, false), params, nil
+}
+
+func buildTraceIDSubquery(q *spanstore.TraceQueryParameters) (string, []interface{}) {
+	quals := make([]string, 0, 15)
```

Comment:
is `quals` self-explanatory? I find it hard to get it without reading the code. Maybe `clauses` suits better.

```suggestion
	clauses := make([]string, 0, 15)
```